### PR TITLE
Fix wallet out-of-sync bug

### PR DESCRIPTION
**Pull request by ChrisSchinnerl**
This PR should fix [#3027](https://github.com/NebulousLabs/Sia/issues/3027).

### CODE REVIEW

File: modules/consensus/consensusset.go [hunk 1]
```diff
@@ -12,7 +12,7 @@ import (
 	"github.com/NebulousLabs/Sia/encoding"
 	"github.com/NebulousLabs/Sia/modules"
 	"github.com/NebulousLabs/Sia/persist"
-	"github.com/NebulousLabs/Sia/sync"
+	siasync "github.com/NebulousLabs/Sia/sync"
```
Comment (DavidVorick) [inline modules/consensus/consensusset.go]: can probably switch this to the more refined NebulousLabs/threadgroup

Comment (DavidVorick) [reply modules/consensus/consensusset.go]: siasync.ThreadGroup should be considered deprecated

File: modules/consensus/subscribe.go [hunk 2]
```diff
@@ -165,16 +168,23 @@ func (cs *ConsensusSet) managedInitializeSubscribe(subscriber modules.ConsensusS
 	})
 	cs.mu.RUnlock()
 	if err != nil {
-		return err
+		return modules.ConsensusChangeID{}, err
+	}
+
+	// Nothing to do if the changeEntry doesn't exist.
+	if !exists {
+		return start, nil
```
Comment (DavidVorick) [inline modules/consensus/subscribe.go]: is this code block reachable? Looks to me like it returns inside the loop if !exists?

File: modules/consensus/consensusset.go (hunk 1 above)
Comment (ChrisSchinnerl) [reply modules/consensus/consensusset.go]: Unfortunately we use `tg.Flush` which is not available in the new package.

File: modules/consensus/subscribe.go (hunk 2 above)
Comment (ChrisSchinnerl) [reply modules/consensus/subscribe.go]: I think so. It's not part of the loop. In line `165` we set `entry, exists = entry.NextEntry(tx)` to get the first entry that we have to send to the new subscriber. If that new entry doesn't exist we return early.

Comment (DavidVorick) [reply modules/consensus/subscribe.go]: ah I see

File: modules/consensus/subscribe.go [hunk 3]
```diff
@@ -165,16 +168,23 @@ func (cs *ConsensusSet) managedInitializeSubscribe(subscriber modules.ConsensusS
...
 	}
 
 	// Send all remaining consensus changes to the subscriber.
+	var latestChangeID modules.ConsensusChangeID
 	for exists {
```
Comment (DavidVorick) [inline modules/consensus/subscribe.go]: at this point, we've already established that `exists` equals true, so we can set the latestChangeID to = `entry.ID()` without needing to worry about correctness. That eliminates any code where we haven't updated the latest change id yet, reducing surface area for mistakes.

File: modules/consensus/accept.go [hunk 4]
```diff
@@ -278,10 +277,11 @@ func (cs *ConsensusSet) managedAcceptBlocks(blocks []types.Block) (blockchainExt
 			if err != nil {
 				return err
 			}
-			// Sanity check - If reverted blocks is zero, applied blocks should also
-			// be zero.
-			if build.DEBUG && len(changeEntry.AppliedBlocks) == 0 && len(changeEntry.RevertedBlocks) != 0 {
-				panic("after adding a change entry, there are no applied blocks but there are reverted blocks")
+			// Sanity check - we should never apply less blocks than we revert.
```
Comment (lukechampine) [inline modules/consensus/accept.go]: fewer

**Pull request merged**